tests/unit/src/Internal/Payment/RouterTest.php

```diff
@@ -0,0 +1,274 @@
+
+	/**
+	 * Tests that the router returns false if a factor is **not present** in the account cache.
+	 */
+	public function test_should_use_new_payment_process_returns_false_with_missing_factor() {
+		$this->mock_db_cache_factors( [] );
+
+		$result = $this->sut->should_use_new_payment_process( [ Factor::USE_SAVED_PM() ] );
+		$this->assertFalse( $result );
+	}
+
+	/**
+	 * Tests that the router returns false if a factor is **false** in the account cache.
+	 */
+	public function test_should_use_new_payment_process_returns_false_with_unavailable_factor() {
+		$this->mock_db_cache_factors( [ Factor::USE_SAVED_PM => false ] );
+
+		$result = $this->sut->should_use_new_payment_process( [ Factor::USE_SAVED_PM() ] );
+		$this->assertFalse( $result );
+	}
+
+	/**
+	 * Tests that the router returns true when a factor is both present, and true in the account cache.
+	 */
+	public function test_should_use_new_payment_process_returns_true_with_available_factor() {
+		$this->mock_db_cache_factors( [ Factor::USE_SAVED_PM => true ] );
+
+		$result = $this->sut->should_use_new_payment_process( [ Factor::USE_SAVED_PM() ] );
+		$this->assertTrue( $result );
+	}
+
+	/**
+	 * Tests that the router handles multiple flags properly,
+	 * and returns false in case any of them is not available.
+	 */
+	public function test_should_use_new_payment_process_with_multiple_factors_returns_false() {
+		$this->mock_db_cache_factors(
+			[
+				Factor::USE_SAVED_PM        => true,
+				Factor::SUBSCRIPTION_SIGNUP => false,
+				Factor::WOOPAY_ENABLED      => true,
+				Factor::PAYMENT_REQUEST     => false,
+			]
+		);
+
+		$result = $this->sut->should_use_new_payment_process(
+			[
+				Factor::USE_SAVED_PM(),
+				Factor::SUBSCRIPTION_SIGNUP(),
+				Factor::WOOPAY_ENABLED(),
+			]
+		);
+		$this->assertFalse( $result );
+	}
+
+	/**
+	 * Tests that the router handles multiple flags properly,
+	 * and returns true when all factors are present.
+	 */
+	public function test_should_use_new_payment_process_with_multiple_factors_returns_true() {
+		$this->mock_db_cache_factors(
+			[
+				Factor::USE_SAVED_PM        => true,
+				Factor::SUBSCRIPTION_SIGNUP => true,
+				Factor::WOOPAY_ENABLED      => true,
+				Factor::PAYMENT_REQUEST     => false,
+			]
+		);
+
+		$result = $this->sut->should_use_new_payment_process(
+			[
+				Factor::USE_SAVED_PM(),
+				Factor::SUBSCRIPTION_SIGNUP(),
+				Factor::WOOPAY_ENABLED(),
+			]
+		);
+		$this->assertTrue( $result );
+	}
+
+	/**
+	 * Check that `get_allowed_factors` returns the factors, provided by the cache.
+	 */
+	public function test_get_allowed_factors_returns_factors() {
+		$cached_factors    = [
+			Factor::SAVE_PM             => true,
+			Factor::SUBSCRIPTION_SIGNUP => false,
+		];
+		$processed_factors = [ Factor::SAVE_PM() ];
+
+		$this->mock_db_cache_factors( $cached_factors, false );
+
+		$result = $this->sut->get_allowed_factors();
+
+		$this->assertIsArray( $result );
+		$this->assertEquals( $processed_factors, $result );
```

Comment:
There are a few cases where because of separate constant instances, I could not use `assertSame` before. Thanks for the suggestion now though, fixed in https://github.com/Automattic/woocommerce-payments/pull/7035/commits/5b1b08b5ad10cb2fe2f7d884fd8274d044b082f5